server/stream.go

```diff
@@ -915,3 +934,166 @@ func (mset *Stream) ackMsg(obs *Consumer, seq uint64) {
 		}
 	}
 }
+
+// Snapshot creates a snapshot for the stream and possibly consumers.
+func (mset *Stream) Snapshot(deadline time.Duration, includeConsumers bool) (*SnapshotResult, error) {
+	mset.mu.Lock()
+	if mset.client == nil || mset.store == nil {
+		mset.mu.Unlock()
+		return nil, fmt.Errorf("invalid stream")
+	}
+	store := mset.store
+	var obs []*Consumer
+	for _, o := range mset.consumers {
+		obs = append(obs, o)
+	}
+	mset.mu.Unlock()
+
+	// Make sure to sync their state.
+	for _, o := range obs {
+		o.writeState()
+	}
+
+	return store.Snapshot(deadline, includeConsumers)
+}
+
+const snapsDir = "__snapshots__"
+
+// RestoreStream will restore a stream from a snapshot.
+func (a *Account) RestoreStream(stream string, r io.Reader) (*Stream, error) {
+	_, jsa, err := a.checkForJetStream()
+	if err != nil {
+		return nil, err
+	}
+
+	sd := path.Join(jsa.storeDir, snapsDir)
+	defer os.RemoveAll(sd)
+
+	if _, err := os.Stat(sd); os.IsNotExist(err) {
+		if err := os.MkdirAll(sd, 0755); err != nil {
+			return nil, fmt.Errorf("could not create snapshots directory - %v", err)
+		}
+	}
+	sdir, err := ioutil.TempDir(sd, "snap-")
+	if err != nil {
+		return nil, err
+	}
+	if _, err := os.Stat(sdir); os.IsNotExist(err) {
+		if err := os.MkdirAll(sdir, 0755); err != nil {
+			return nil, fmt.Errorf("could not create snapshots directory - %v", err)
+		}
+	}
+
+	gzr, err := gzip.NewReader(r)
+	if err != nil {
+		return nil, err
+	}
+	defer gzr.Close()
+	tr := tar.NewReader(gzr)
+
+	for {
+		hdr, err := tr.Next()
+		if err == io.EOF {
+			break // End of snapshot
+		}
+		if err != nil {
+			return nil, err
+		}
+		fpath := path.Join(sdir, filepath.Clean(hdr.Name))
+		pdir := filepath.Dir(fpath)
+		os.MkdirAll(pdir, 0755)
```

Comment:
I know we use `755` for all the dirs now so its nothing new, but probably `750` be better?

Reply:
ok will update.